[framework] unify list of countries across domains

grossmannmartin opened the pull request:
| Q             | A
| ------------- | ---
|Description, reason for the PR| Unify countries across domains, so country on one domain is the same entity as on the others. 
|New feature| No <!-- Do not forget to update docs/ -->
|BC breaks| Yes <!-- Do not forget to update UPGRADE.md -->
|Fixes issues| ... <!-- Write "closes #123" for the issue to be closed automatically during merge -->
|Standards and tests pass| Yes
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

- now there is only one list of countries for all domains.
- countries cannot be deleted
- countries are sorted by priority and by name if same priority occurs
- each country has
  - name in all languages
  - ISO compliant code
  - visibility on domain
  - priority on each domain
- customer can choose only countries allowed on domain

TomasLudvik commented on packages/framework/src/Migrations/DataModifiers/CountryDataModifierVersion20190121094400.php:
```diff
@@ -0,0 +1,192 @@
+    public function getNewIdCodePair(): array
+    {
+        $tmp = [];
+        foreach ($this->data as $row) {
```
This function is too complex, data here should be ordered by domain_id so you could remove first foreach and remover $row['domain_id] === 1 from condition

grossmannmartin replied:
I simplified the function. Please check if it's ok.

TomasLudvik commented on packages/framework/src/Form/Admin/Country/CountryFormType.php:
```diff
@@ -30,19 +29,12 @@ class CountryFormType extends AbstractType
      */
     protected $countryFacade;
 
-    /**
```
This should not be part of the last commit

grossmannmartin replied:
You're absolutely right. I noticed it while writing the test. I can move it on rebase.

grossmannmartin commented on packages/framework/src/Form/Admin/Country/CountryFormType.php:
```diff
@@ -2,48 +2,133 @@
-            'data_class' => CountryData::class,
-            'attr' => ['novalidate' => 'novalidate'],
-        ]);
+        if ($this->countryCode === null || $countryCodeValue !== $this->countryCode) {
```
You're right :+1:
I fixed it and added a test for this use case.

grossmannmartin commented on packages/framework/src/Migrations/Version20190121094400.php:
```diff
@@ -0,0 +1,61 @@
+            ADD
+                CONSTRAINT FK_4537E7F0F92F3E70 FOREIGN KEY (country_id) REFERENCES countries (id) NOT DEFERRABLE INITIALLY IMMEDIATE');
+
+        $this->sql('ALTER TABLE countries ALTER COLUMN name DROP NOT NULL');
```
Actually, I have `priority` in `CountryDomain`entity defined with default value too, so dropping default result in inconsistent schema and metadata.

So you suggest to remove default value completely and keep setting the value only in the constructor of the entity?

TomasLudvik replied:
Well yes.

grossmannmartin commented on packages/framework/src/Model/Country/CountryData.php:
```diff
@@ -5,7 +5,7 @@
-     * @var string
+     * @var string[]
      */
     public $name;
```
:+1: fixed

grossmannmartin commented on packages/framework/src/Model/Country/CountryFacade.php:
```diff
@@ -86,16 +85,27 @@ public function edit($countryId, CountryData $countryData)
+     * @param string $countryCode
+     * @return \Shopsys\FrameworkBundle\Model\Country\Country|null
+     */
+    public function getByCode(string $countryCode): ?Country
```
Right! Fixed

grossmannmartin commented on packages/framework/src/Migrations/DataModifiers/CountryDataModifierVersion20190121094400.php:
```diff
@@ -0,0 +1,192 @@
+
+namespace Shopsys\FrameworkBundle\Migrations\DataModifiers;
+
+use function array_diff;
```
It's optimization for parser. With native functions, you should use either `\array_diff(` or import the function as it is here. 
By default parser looking for the function in a namespace and after that in a global scope. With import or FQN you force the parser to look in a global scope first. 
It's considered a good practice.

And because I don't like backward slashes in a front of things in code, I choose this way.

TomasLudvik replied:
Ok, thanks for the explanation. I would love if you could create user story/issue or at least discuss it on our internal meeting so we decide if we want to use it. 

It would be also nice to know some performance impact. 

Until then please remove this code.

grossmannmartin commented on packages/framework/src/Resources/views/Admin/Content/Country/list.html.twig:
```diff
@@ -1,9 +1,21 @@
-{% endblock %}
\ No newline at end of file
+
+    {% embed '@ShopsysFramework/Admin/Inline/FixedBar/fixedBar.html.twig' %}
```
Fixed!

grossmannmartin commented on packages/framework/src/Model/Country/Exception/CountryDomainNotFoundException.php:
```diff
@@ -0,0 +1,23 @@
+    public function __construct(int $domainId, int $countryId = null, Exception $previous = null)
+    {
+        $countryDescription = $countryId !== null ? sprintf('with ID %d', $countryId) : 'without ID';
+        $message = sprintf('TransportDomain for transport %s and domain ID %d not found.', $countryDescription, $domainId);
```
Oh, copypaste error. Fixed.

grossmannmartin commented on packages/framework/src/Model/Country/CountryDataFactory.php:
```diff
@@ -28,9 +46,32 @@ public function createFromCountry(Country $country): CountryData
+    protected function fillNew(CountryData $countryData): void
+    {
+        foreach ($this->domain->getAllIds() as $domainId) {
+            $countryData->enabled[$domainId] = true;
```
I discussed it with @LukasHeinz and he prefers to have all checkbox checked (default state is taken from here).
If it were false, it would complicate DataFixtures too (need to manually enable countries).

Do you agree that it stays true?

TomasLudvik replied:
Yes.

TomasLudvik commented on packages/framework/tests/Unit/Form/Admin/Country/CountryFormTypeTest.php:
```diff
@@ -141,10 +178,11 @@ protected function getExtensions(): array
      * @return \Symfony\Component\Form\FormInterface
      */
-    private function createCountryForm(): FormInterface
+    private function createCountryForm(?Country $country = null): FormInterface
```
Please remove "?" before country, it is redundant.

grossmannmartin commented:
Hi @TomasLudvik,
based on insights from testing, I have added another commit. 

If someone places an order with some country and the country is disabled on the domain afterward, the first country is selected in detail of order in admin. That may cause that mere re-saving of order changes data and administrator may not notice at all.

I changed a code a bit, so in the detail of the order are shown always all created countries.

Country select is present yet in a customer detail (shipping and billing address), but there I choose not to show all the countries. Selecting disabled country cannot work on frontend properly, so it doesn't make sense here.

Can I propose a review of the last commit? And some thoughts about it maybe.
